### PR TITLE
Add actions to enable/disable coordinape user, in Admin Page

### LINKED ISSUE #558: "Coordinape In Your Circle": ENABLE/DISABLE in members list in admin view
**Problem**
We've added a Coordinape user in all circles to enable people to give to Coordinape if they feel inspired as a token of thanks or just to support Coordinape. The problem is that we have not given our Coordinape Humans enough context as to what this is all about and many admins have simply deleted it. Others have ignored. A few have given to Coordinape! 

**Specification**
- Add the ability to "Enable" or "Disable" the Coordinape member if you are an admin. 
- Add tooltip next to ENABLE/DISABLE explaining the option. 

This issue is a little more complex since it may involve some deeper changes to implement the state of "Coordinape as a Circle Member". 

For the enable, this would require a new hasura mutation that restores the user, mutation would set`deleted_at` back to a `null` value. This mutation can be for generic use in restoring normal users.
(validate that the address is still unique within the circle if restored).

For the disable you can use the hasura mutation that deletes a normal user.

Design: 
https://www.figma.com/file/Db8KthuZRjT5CfMvZsjTT2/Coordinape-Design-v7?node-id=2692%3A9277

### PR DESCRIPTION
**Pull request by amrro**
Fixes #558 #559
____

In order to reinsstate the coordinape user, this creates a dummy coorindape user object and adds to the table in case of: 1) doesn't exist at all, 2) soft-deleted.
In case there are no users yet in the circle, coordinape user won't be shown.

Since `createUser` cannot add user with a `role` that is larger than 1, I used the coordinape's address as indication if this is the coordinape user or not.

Screen Shot 2022-06-16 at 12 38 11 AM

____

### Caveats
- In the seeding data, Coordinape doesn't exist in any circle. Make sure to add it manually beforehand. You can run: 
```gql
mutation coordinape {
  createUser(payload: {address: "0xfad763da9051953fea58c2304395719a3b6ba361", circle_id: 33, name: "Coordinape"}) {
    UserResponse {
      address
      name
    }
  }
}
```
Then update its `role` in the database manuallay to value `2`.

### CODE REVIEW

File: src/pages/AdminPage/AdminPage.tsx [hunk 1]
```diff
@@ -201,6 +207,25 @@ const AdminPage = () => {
             : undefined
         }
       />
+      <ActionDialog
```
Comment (levity) [inline src/pages/AdminPage/AdminPage.tsx:210]: you don't have to use this here; a simple `window.confirm` works as well

**Review by levity — CHANGES_REQUESTED** [review]
Thanks for this, @amrro. Please note the spec in #558:

> For the enable, this would require a new hasura mutation that restores the user, mutation would set `deleted_at` back to a null value. 

It's important that the special role for the coordinape user is correctly set. Creating a user with the same address but the role of a normal user is going to confuse other code and analytics, so it's not an acceptable solution. Also, we don't want to create multiple deleted users with the same address if the Coordinape user is enabled & disabled multiple times.

Instead, please create a new mutation that searches for a user with non-null deleted_at and role = 2, and un-deletes them by setting deleted_at to null. All new circles have a Coordinape user created by default, so you can assume one always exists in the circle that you can undelete.

File: src/lib/zod/index.ts [hunk 2]
```diff
@@ -258,6 +258,12 @@ export const updateCircleInput = z
   })
   .strict();
 
+export const restoreCoordinapeInput = z
```
Comment (levity) [inline src/lib/zod/index.ts:261]: @topocount @zashton IMO these should be co-located with handlers when they're not used in multiple places; what do you think?

File: src/pages/AdminPage/components.tsx [hunk 3]
```diff
@@ -832,23 +890,33 @@ export const ContributorsTable = ({
...
                       area="wide"
                       align="left"
+                      disabled={u.deleted_at !== null}
                     >
                       <UserName user={u} />
                     </Table.Cell>
 
-                    <Table.Cell key={`address-${u.id}`}>
+                    <Table.Cell
+                      key={`address-${u.id}`}
```
Comment (levity) [inline src/pages/AdminPage/components.tsx:899]: i overlooked this in the previous review... all these `key` properties on Cell are unnecessary. Only the top-level `Row` tag needs one

File: src/lib/gql/mutations.ts [hunk 4]
```diff
@@ -415,6 +415,26 @@ export async function updateTeammates(circleId: number, teammates: number[]) {
...
+            circle_id: circleId,
+          },
+        },
+        {
+          success: true,
+        },
+      ],
+    },
+    {
+      operationName: 'resotre_coordinape',
```
Comment (levity) [inline src/lib/gql/mutations.ts:433]:
```suggestion
      operationName: 'restore_coordinape',
```

File: src/lib/gql/mutations.ts [hunk 5]
```diff
@@ -415,6 +415,26 @@ export async function updateTeammates(circleId: number, teammates: number[]) {
   return updateTeammates;
 }
 
+export async function restoreCoordinpeUser(circleId: number) {
```
Comment (levity) [inline src/lib/gql/mutations.ts:418]:
```suggestion
export async function restoreCoordinapeUser(circleId: number) {
```

File: src/hooks/useApiAdminCircle.ts [hunk 6]
```diff
@@ -97,16 +97,24 @@ export const useApiAdminCircle = (circleId: number) => {
     [circleId]
   );
 
+  const restoreCoordinape = useRecoilLoadCatch(
+    () => async (circleId: number) => {
+      await mutations.restoreCoordinpeUser(circleId);
```
Comment (levity) [inline src/hooks/useApiAdminCircle.ts:102]:
```suggestion
      await mutations.restoreCoordinapeUser(circleId);
```

File: hasura/metadata/actions.yaml [hunk 7]
```diff
@@ -112,6 +112,15 @@ actions:
       value_from_env: HASURA_EVENT_SECRET
   permissions:
   - role: user
+- name: restoreCoordinape
+  definition:
+    kind: synchronous
+    handler: '{{HASURA_API_BASE_URL}}/actions/restoreCoordinape'
```
Comment (levity) [inline hasura/metadata/actions.yaml:118]: we recently changed the way we add handlers (see #1046 for context); please follow the new pattern

File: src/pages/AdminPage/components.tsx [hunk 8]
```diff
@@ -887,14 +964,28 @@ export const ContributorsTable = ({
...
-                          ? () => setDeleteUserDialog(u)
-                          : undefined
-                      )}
+                      {u.role === USER_ROLE_COORDINAPE
+                        ? renderCoordinapeActions(u.deleted_at === null, () => {
+                            const shouldEnable = u.deleted_at !== null;
+                            const confirm = window.confirm(
+                              `${
+                                shouldEnable ? 'Enable' : 'Disable'
+                              } Coordinape in this ${circle.name}?`
```
Comment (levity) [inline src/pages/AdminPage/components.tsx:973]:
```suggestion
                              } Coordinape in this circle?`
```

File: hasura/metadata/actions.graphql [hunk 9]
```diff
@@ -103,7 +107,6 @@ input CreateUserInput {
   non_receiver: Boolean
   role: Int
   starting_tokens: Int
```
Comment (levity) [inline hasura/metadata/actions.graphql:109]:
```suggestion
  starting_tokens: Int
  fixed_payment_amount: Float
```

File: hasura/metadata/actions.graphql [hunk 10]
```diff
@@ -125,7 +128,6 @@ input AdminUpdateUserInput {
   fixed_non_receiver: Boolean
   non_receiver: Boolean
   role: Int
```
Comment (levity) [inline hasura/metadata/actions.graphql:130]:
```suggestion
  role: Int
  fixed_payment_amount: Float
```

File: src/pages/AdminPage/components.tsx [hunk 11]
```diff
@@ -291,19 +295,64 @@ export const TableLink = styled(RouterLink, {
...
+        flexDirection: 'column',
+      }}
+    >
+      <Text size="large" css={{ my: '$md' }}>
+        Why is Coordinape in your circle?
+      </Text>
+      <p>
+        We&apos;re experimenting with the gift circle mechanism as our revenue
+        model. By default, Coordinape appears in your circle and any user can
+        allocate to Coordinape. To remove the Coordinape user, click the trash
```
Comment (levity) [inline src/pages/AdminPage/components.tsx:313]: Change this from "To remove..." to "To show or hide the Coordinape user, use the links on the right side of this row."

File: src/pages/AdminPage/components.tsx [hunk 12]
```diff
@@ -832,23 +890,33 @@ export const ContributorsTable = ({
...
                       align="left"
+                      disabled={u.deleted_at !== null}
                     >
                       <UserName user={u} />
                     </Table.Cell>
 
-                    <Table.Cell key={`address-${u.id}`}>
+                    <Table.Cell
+                      key={`address-${u.id}`}
+                      disabled={u.deleted_at !== null}
```
Comment (levity) [inline src/pages/AdminPage/components.tsx:900]: I'm bikeshedding here... but you could have avoided adding all these `disabled` lines and put it in the Row tag once instead.

**Review by levity — CHANGES_REQUESTED** [review]

File: src/pages/AdminPage/components.tsx (hunk 12 above)
Comment (amrro) [reply src/pages/AdminPage/components.tsx:900]: Pardon me. I tried to add it to the whole row. But we have a single column that is not disabled. And a child cannot have higher opacity than its parents, and workaround would make it complex.

Comment (levity) [reply src/pages/AdminPage/components.tsx:900]: Ah, sorry, I overlooked the non-disabled cell. How about something like:

```
<Row css={isCoordinapeUser ? {
  '> td': { opacity: 0.3 },
  '> td.normal': { opacity: 1 }
}: null}>
  <Cell>a</Cell>
  <Cell className="normal">b</Cell>
</Row>
```

**Comment by spydermonkee** [comment]
@amrro - Hey there!
What's the latest status on this PR? Is it still waiting for review, or are there still requested changes that need to be addressed?

File: api/hasura/actions/_handlers/restoreCoordinape.ts [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+      update_users_by_pk: [
+        {
+          pk_columns: { id: existingCoordinape.id },
+          _set: { deleted_at: null },
+        },
+        { __typename: true },
+      ],
+    },
+    {
+      operationName: 'resotre_coordinape',
```
Comment (levity) [inline api/hasura/actions/_handlers/restoreCoordinape.ts:55]:
```suggestion
      operationName: 'restore_coordinape',
```

**Comment by levity** [comment]
> @amrro - Hey there! What's the latest status on this PR? Is it still waiting for review, or are there still requested changes that need to be addressed?

I'll review this now

**Comment by levity** [comment]
I've merged this into a copy of your branch in our repo, so that we can run CI on it. (A fix so that CI can run on fork branches is coming soon.)

**Pull request merged**